### DISCUSSION TITLE
Super() on extended/overridden class

### DISCUSSION THREAD

**Discussion by Chetral** [evennia/evennia#3553] [Evennia Questions] [answered]
Hello.
I'm creating a dungeon using the Dungeon model in Tutorial ([dungeon.py](https://github.com/evennia/evennia/blob/main/evennia/contrib/tutorials/evadventure/dungeon.py)).
Previously I had overridden the Exit class `at_traverse()` hook as i needed to change some defaults (not "leaving towards [room name], but leaving "towards [direction]", translation of messages in italian, movement stamina, delay in moving).
Note that the `typeclasses.exits.Exit.at_taverse()` do not have any super() recall, i've overridden it completely.

Following the documentation i imported my Exit definition in dungeon.py and made EvAdventureDungeonExit inherit from it:
`from typeclasses.exits import Exit
[...]
class EvAdventureDungeonExit(Exit):
`
When i move in the dungeon i get the default english messages!

I've searched on the web and found out that `super().at_traverse(traversing_object, target_location, **kwargs)` in the EvAdventureDungeonExit will actually call the original parent definition.
So i changed the line in `super(EvAdventureDungeonExit, self).at_traverse(traversing_object, target_location, **kwargs)`, according to an answer in StackOverflow this will call the at_traverse hook of the class last class definition before the one called.  Since the EvAdventureDungeonExit inherits directly from my typeclasses.exits.Exit and in my Exit at_traverse hook I don't have a super() call I expected this to work... but i still get default messages when a character or a mob moves around.

Any suggestion?

**Comment by Griatch** [comment]
First of all, you can forget about `super(EvadventureDungeonExit, self)` form for this case, that's something that can come into play for multiple-inheritance which is not a thing here. It was required to put on this form in Python2, not so for more recent version of Python. 

As far as I understand, this is your setup: 

```py
# in mygame/typeclasses/exits.py

class Exit(DefaultExit)::
     # modified class 

     def at_traverse(self):
         # changes made for italian

# in some other module in gamedir
class EvAdventureDungonExit(Exit):
     # unchanged class from the contrib except the inheritance
      
    def at_traverse(self):
        # ...
        super().at_traverse()

```
This looks correct, and you should be getting your modified `Exit.at_traverse` after the custom dungeon code in the dungeon exit ran. 

> Question  Are you modifying the `EvAdventureDungeonExit` directly in core Evennia? If you do that we can't really help you, you are on your own. You should always apply modifications by inheriting to your own code in your game dir. Copy things out of the tutorial and rename classes appropriately.

Otherwise, what you should check is that the exits you create are _actually_ your custom exits and not the originals. Analyze their typeclass in-game and see that they really reference the relevant overridden version of the dungeon-exit. Also, actual English text is found in other methods called by `at_traverse`, make sure these are not overridden.

**Reply by Chetral** [reply]
Hi Griatch,
the setup you posted is correct. I didn't mention that i use my `typeclasses.extis.Exit` in other places and it works properly, with the italian messages and so on.

I copied the whole dungeon.py file in my "world" folder as i'm modifying it to spawn mobs and objects. Actually I also renamed all "EvAdventure" in "fogne" ("sewers" in italian). I checked the exits with EXAMINE Direction and they are FogneExit as expected.

I'm also using my version of Room for inheritance (instead of `from .rooms import EvAdventureRoom` i import `from typeclasses.rooms import Room`

**Comment by Griatch** [comment]
On the Exit object is the `.exit_command` property that points to the `ExitCommand` being used to handle traversals. This _should_ just work normally with your overridden exit type (it only calls methods on the object it's associated with), but it could be worth to double-check, so that it doesn't end up calling methods on the wrong object for some reason. 

Otherwise, from what you describe, things should work; you can debug further by putting `print` statements in key methods to see what is actually being called - most likely there's something not overridden properly somewhere. You can also use the debugger (search to docs for `debugging` or 'debugger' and you'll find the instructions) to step through what happens when you try to traverse an exit.

**Reply by Chetral** [reply]
Thanks for the hints. I will check the `.exit_command` and the doc for debugging, will return with (hopefully) the problem solved :)

**Comment by Chetral** [comment, marked as answer]
Found it!
It was a (my) error in the wondering mobiles, calling the wrong hook when the exit was NOT a wilderness exit.

**Reply by Griatch** [reply]
Glad you found it! :+1:
